src/Sarif/Writers/ResultLogJsonWriter.cs

```diff
@@ -167,14 +216,36 @@ public void Dispose()
                 _jsonWriter.WriteEndObject(); // End: resultsLog
             }
 
-            _writeState = State.Disposed;
+            _writeConditions |= Conditions.Disposed;
+        }
+
+        private void EnsureInitialized()
+        {
+            if (_writeConditions == Conditions.None)
+            {
+                Initialize();
+            }
+        }
+
+        private void EnsureStateNotAlreadySet(Conditions invalidConditions)
+        {
+            Conditions observedInvalidConditions = _writeConditions & invalidConditions;
+            if (observedInvalidConditions != Conditions.None)
+            {
+                // 	InvalidState	One or more invalid states were detected during serialization: {0}	
+                throw new InvalidOperationException(string.Format(SarifResources.InvalidState, observedInvalidConditions));
```

Comment:
Now I also understand why you changed the expected exceptions in some of your tests. You had this general purpose method that _always throws the same exception_. So you didn't get the exception you really wanted. With my proposed design, you get `ObjectDisposedException` when you should, and `InvalidOperationException` when you should.